### ISSUE TITLE
Relicense under dual MIT/Apache-2.0

### ISSUE THREAD

**Issue by emberian** [vlad003/rust-oath#3]
# Why?

The MIT license requires reproducing countless copies of the same copyright
header with different names in the copyright field, for every MIT library in
use. The Apache license does not have this drawback, and has protections from
patent trolls and an explicit contribution licensing clause. However, the
Apache license is incompatible with GPLv2. This is why Rust is dual-licensed as
MIT/Apache (the "primary" license being Apache, MIT only for GPLv2 compat), and
doing so would be wise for this project. This also makes this crate suitable
for inclusion in the Rust standard distribution and other project using dual
MIT/Apache.
# How?

To do this, get explicit approval from each contributor of copyrightable work
(as not all contributions qualify for copyright) and then add the following to
your README:

```
## License

Licensed under either of
 * Apache License, Version 2.0 ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)
at your option.

### Contribution

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you shall be dual licensed as above, without any
additional terms or conditions.
```

and in your license headers, use the following boilerplate (based on that used in Rust):

```
// Copyright (c) 2015 t developers
// Licensed under the Apache License, Version 2.0
// <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT
// license <LICENSE-MIT or http://opensource.org/licenses/MIT>,
// at your option. All files in the project carrying such
// notice may not be copied, modified, or distributed except
// according to those terms.
```

And don't forget to update the `license` metadata in your `Cargo.toml`!
# Contributor checkoff
- [ ] @vlad003

**Comment by avacariu** [comment]
1. The "Why?" section is not very convincing. Even if you're somehow using "countless" libraries, I disagree that having to reproduce the same header multiple times is a significant drawback. In fact, the Apache license requires that _"You must retain, in the Source form of any Derivative Works that You distribute, all copyright, patent, trademark, and attribution notices from the Source form of the Work, excluding those notices that do not pertain to any part of the Derivative Works"_, so I don't see how that's any better.
2. If this was a genuine issue raised by an actual user, and was not automated, I'd consider it, but as it stands, I refuse.

**Comment by emberian** [comment]
1. That only applies to source distributions, not binary distributions. The primary motivation is not the header duplication, but the code sharing possible with the Rust standard distribution and other MIT/Apache-2.0 projects.
2. I only automated this because I do genuinely care about software licensing, and the Rust ecosystem is largely dual MIT/Apache-2.0. My ulterior motive is making as many crates as possible seamlessly compatible for shipping with the [Robigalia project](https://robigalia.org). There are many Apache xor MIT crates that are that way for no particular reason.

Of course, it's your code. Feel free to close the issue if you just don't want to do it!

**Comment by avacariu** [comment]
The MIT license doesn't conflict with either of those uses, so I'll stick with a single license.

**Comment by emberian** [comment]
This crate can't be included in dual MIT/Apache-2.0 aggregations because then users can't chose to use the Apache-2.0, only MIT, or separate licensing for each component.
